flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java

```diff
@@ -0,0 +1,86 @@
+import org.apache.flink.runtime.state.StateUtil;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Delegate class responsible for checkpoints cleaning and counting the number of checkpoints yet
+ * to clean.
+ */
+public class CheckpointsCleaner {
+	/** The executor used for asynchronous calls, like potentially blocking I/O. */
+	private final Executor executor;
+	AtomicInteger numberOfCheckpointsToClean;
+
+	Runnable checkpointCleaningFinishedCallback;
+
+	private static final Logger LOG = LoggerFactory.getLogger(CheckpointsCleaner.class);
+
+	public CheckpointsCleaner(Executor executor) {
+		this.executor = executor;
+		this.numberOfCheckpointsToClean = new AtomicInteger(0);
+	}
+
+	public void setCheckpointCleaningFinishedCallback(Runnable checkpointCleaningFinishedCallback) {
+		this.checkpointCleaningFinishedCallback = checkpointCleaningFinishedCallback;
+	}
+
+	public int getNumberOfCheckpointsToClean() {
+		return numberOfCheckpointsToClean.get();
+	}
+
+	/**
+	 * Asynchronously call a discard callback on on the ioExecutor
+	 * (FixedThreadPool of configurable size of default 4*CPU cores)
+	 * and count the number of checkpoints that are waiting to clean.
+	 * @param completedCheckpoint the checkpoint to discard
+	 * @param discardCallback the discard callback to call
+	 */
+	public void asyncDiscardCheckpointAndCountCheckpoints(CompletedCheckpoint completedCheckpoint, ThrowingConsumer<CompletedCheckpoint, Exception> discardCallback){
+		numberOfCheckpointsToClean.incrementAndGet();
+		executor.execute(() -> {
+			try {
+				discardCallback.accept(completedCheckpoint);
+			} catch (Exception e) {
+				LOG.warn("Could not discard completed checkpoint {}.", completedCheckpoint.getCheckpointID(), e);
+			}
+			finally {
+				numberOfCheckpointsToClean.decrementAndGet();
+				Preconditions.checkNotNull(checkpointCleaningFinishedCallback);
+				checkpointCleaningFinishedCallback.run();
+			}
+		});
+	}
+
+	public void asyncDiscardPrivateStatesAndCountCheckpoints(
+		Map<OperatorID, OperatorState> operatorStates, CheckpointStorageLocation targetLocation,
+		long checkpointId, JobID jobId) {
+		numberOfCheckpointsToClean.incrementAndGet();
+		executor.execute(() -> {
+			// discard the private states.
+			// unregistered shared states are still considered private at this point.
+			try {
+				StateUtil.bestEffortDiscardAllStateObjects(operatorStates.values());
+				targetLocation.disposeOnFailure();
+			} catch (Throwable t) {
+				LOG.warn(
+					"Could not properly dispose the private states in the pending checkpoint {} of job {}.",
+					checkpointId, jobId, t);
+			} finally {
+				operatorStates.clear();
+				numberOfCheckpointsToClean.decrementAndGet();
```

Comment:
this method is called in `PendingCP.dispose()` which is called in `PendingCP.finalizeCheckpoint()` which is called when CP are completed. so if I do so, a new CP trigger request will be picked when a given CP is completed and then when it is discarded. So 2 more CP requests for one single checkpoint. Should I make that the counter logic is called only in case of failure (`PendingCP.abort()`) to avoid that case ?

Reply:
`PendingCP.finalizeCheckpoint()` calls `PendingCP.dispose()` with `releaseState=false`, so this method should only run for aborts.
And if it ever changes, I don't think it's a problem to check the queue one more time.

Reply:
ok thanks for this, makes sense

Reply:
done